### ISSUE TITLE
Assertion `Info && "Invalid name for a VFABI variant."' failed.

### ISSUE THREAD

**Issue by regehr** [llvm/llvm-project#71892]
this test is a reduced version of `llvm/test/Transforms/LoopVectorize/scalarize-masked-call.ll`

```llvm
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"

define void @cond_call() {
entry:
  br label %for.body

for.body:                                         ; preds = %for.body, %entry
  %foo.ret = call i64 @foo()
  br label %for.body
}

declare i64 @foo() #0

attributes #0 = { "vector-function-abi-variant"="_ZGV_LLVM_M4v_foo(vector_foo)" }
```

I get:

```
regehr@ohm:~/reduce$ /home/regehr/llvm-project/for-alive/bin/opt -O2 reduced.ll -o /dev/null
opt: /home/regehr/llvm-project/llvm/lib/Analysis/VectorUtils.cpp:1473: void llvm::VFABI::getVectorVariantNames(const llvm::CallInst&, llvm::SmallVectorImpl<std::__cxx11::basic_string<char> >&): Assertion `Info && "Invalid name for a VFABI variant."' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /home/regehr/llvm-project/for-alive/bin/opt -O2 reduced.ll -o /dev/null
 #0 0x00007fd3a2ffe950 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMSupport.so.18git+0x1fe950)
 #1 0x00007fd3a2ffb95f llvm::sys::RunSignalHandlers() (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMSupport.so.18git+0x1fb95f)
 #2 0x00007fd3a2ffbab5 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007fd3a2642520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007fd3a26969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007fd3a26969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007fd3a26969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x00007fd3a2642476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007fd3a26287f3 abort ./stdlib/abort.c:81:7
 #9 0x00007fd3a262871b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x00007fd3a2639e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#11 0x00007fd3a3d060bf llvm::VFABI::getVectorVariantNames(llvm::CallInst const&, llvm::SmallVectorImpl<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>>&) (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMAnalysis.so.18git+0x5060bf)
#12 0x00007fd3a585f086 llvm::VFDatabase::getVFABIMappings(llvm::CallInst const&, llvm::SmallVectorImpl<llvm::VFInfo>&) (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMVectorize.so.18git+0x5f086)
#13 0x00007fd3a586560a llvm::LoopVectorizationLegality::canVectorizeInstrs() (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMVectorize.so.18git+0x6560a)
#14 0x00007fd3a58667e7 llvm::LoopVectorizationLegality::canVectorize(bool) (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMVectorize.so.18git+0x667e7)
#15 0x00007fd3a58c8529 llvm::LoopVectorizePass::processLoop(llvm::Loop*) (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMVectorize.so.18git+0xc8529)
#16 0x00007fd3a58cc771 llvm::LoopVectorizePass::runImpl(llvm::Function&, llvm::ScalarEvolution&, llvm::LoopInfo&, llvm::TargetTransformInfo&, llvm::DominatorTree&, llvm::BlockFrequencyInfo*, llvm::TargetLibraryInfo*, llvm::DemandedBits&, llvm::AssumptionCache&, llvm::LoopAccessInfoManager&, llvm::OptimizationRemarkEmitter&, llvm::ProfileSummaryInfo*) (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMVectorize.so.18git+0xcc771)
#17 0x00007fd3a58cd670 llvm::LoopVectorizePass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMVectorize.so.18git+0xcd670)
#18 0x00007fd3a62b7ec6 llvm::detail::PassModel<llvm::Function, llvm::LoopVectorizePass, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMPasses.so.18git+0xb7ec6)
#19 0x00007fd3aa782fb1 llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMAMDGPUCodeGen.so.18git+0x382fb1)
#20 0x00007fd3a3525647 llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMCore.so.18git+0x325647)
#21 0x00007fd3aa762746 llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMAMDGPUCodeGen.so.18git+0x362746)
#22 0x00007fd3a35236c1 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/regehr/llvm-project/for-alive/bin/../lib/libLLVMCore.so.18git+0x3236c1)
#23 0x0000557f58a33034 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (/home/regehr/llvm-project/for-alive/bin/opt+0x31034)
#24 0x0000557f58a2109c main (/home/regehr/llvm-project/for-alive/bin/opt+0x1f09c)
#25 0x00007fd3a2629d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#26 0x00007fd3a2629e40 call_init ./csu/../csu/libc-start.c:128:20
#27 0x00007fd3a2629e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#28 0x0000557f58a24f95 _start (/home/regehr/llvm-project/for-alive/bin/opt+0x22f95)
Aborted (core dumped)
regehr@ohm:~/reduce$ 
```

**Comment by fhahn** [comment]
@huntergr-arm I think you have been working in this area, could you take a look?

**Comment by huntergr-arm** [comment]
The crash is due to `vector_foo` not being defined in the module, though there's also a problem in the mangled name -- `4v` implies that there's a vector argument to the function variant, but foo doesn't take any arguments.

I'm taking over the work started in https://github.com/llvm/llvm-project/pull/67430 (though I don't think I can actually take over that pull request, so I'll just create a new one) which can address this case, though it wouldn't warn about the problems. 

Do we want to extend the IR verifier to complain about such problems, or is it sufficient to just not try vectorizing with an invalid mangled name or missing declaration for the vector variant?

**Comment by fhahn** [comment]
> Do we want to extend the IR verifier to complain about such problems, or is it sufficient to just not try vectorizing with an invalid mangled name or missing declaration for the vector variant?

AFAICT the documentation for the attribute is part of LangRef (https://llvm.org/docs/LangRef.html#call-site-attributes), so IMO I would make sense to check this in the IR verifier.

**Comment by huntergr-arm** [comment]
https://github.com/llvm/llvm-project/pull/72260 should address the crash once it lands; I'll improve the verifier in a separate PR since that will require at least some of the VFABI code to move into LLVM core (otherwise the core .so will need to link against the analysis .so).
